apps/aecore/lib/aecore/util/state_trees.ex

```diff
@@ -0,0 +1,95 @@
+defmodule Aecore.Util.StateTrees do
+  alias Aecore.Account.AccountStateTree
+  alias Aecore.Contract.CallStateTree
+  alias Aecore.Contract.ContractStateTree
+  alias Aecore.Channel.ChannelStateTree
+  alias Aecore.Naming.NamingStateTree
+
+  @moduledoc """
+  Abstract module defining functions for known Tree types
+  """
+  defmacro __using__(_) do
+    quote location: :keep do
+      alias Aecore.Contract.{Contract, ContractStateTree}
+      alias Aecore.Chain.Identifier
+      alias Aecore.Naming.{Name, NameCommitment}
+      alias Aecore.Util.StateTrees
+      alias Aeutil.PatriciaMerkleTree
+      alias Aeutil.Serialization
+      alias MerklePatriciaTree.Trie
+
+      @typedoc "Hash of the tree"
+      @type hash :: binary()
+
+      @spec init_empty :: Trie.t()
+      def init_empty do
+        PatriciaMerkleTree.new(StateTrees.tree_type(__MODULE__))
+      end
+
+      @spec put(Trie.t(), binary(), map()) :: Trie.t()
+      def put(tree, key, value) do
+        serialized_state = Serialization.rlp_encode(value)
+        PatriciaMerkleTree.enter(tree, key, serialized_state)
+      end
+
+      @spec get(Trie.t(), binary()) :: :none | map()
+      def get(tree, key) do
+        case PatriciaMerkleTree.lookup(tree, key) do
+          {:ok, serialized_value} ->
+            {:ok, deserialized_value} = Serialization.rlp_decode_anything(serialized_value)
+
+            case deserialized_value do
+              %Name{} ->
+                hash = Identifier.create_identity(key, :name)
+                %Name{deserialized_value | hash: hash}
+
+              %NameCommitment{} ->
+                hash = Identifier.create_identity(key, :commitment)
+                %NameCommitment{deserialized_value | hash: hash}
+
+              %Contract{} ->
+                identified_id = Identifier.create_identity(key, :contract)
+                store_id = Contract.store_id(%{deserialized_value | id: identified_id})
+
+                %Contract{
+                  deserialized_value
+                  | id: identified_id,
+                    store: ContractStateTree.get_store(store_id, tree)
+                }
+
+              _ ->
```

Comment:
do we use the default here? otherwise an error message would be better, or not handling the case at all

Reply:
We are using `case do` because of the identifiers(in the structures that we need), as epoch does. Where we don't need identifiers we just return the deserialized value.

Reply:
ah okay!